File: lim/연결리스트/Doubly-LinkedList - 코드트리 유치원.py
import sys

next_id = 1
node_by_id = dict()


class Node:
    def __init__(self):
        global next_id
        global node_by_id
        self.value = next_id
        self.prev = None
        self.next = None
        next_id += 1

        node_by_id[self.value] = self


def sol(queries):
    global root
    root = Node()

    def create_new_line(size):
        start_node = Node()
        size -= 1

        end_node = start_node
        while size > 0:
            node = Node()
            end_node.next = node
            node.prev = end_node
            end_node = node
            size -= 1

        return start_node, end_node

    def get_by_id(target_id):
        global node_by_id
        return node_by_id[target_id]

    def insert_next(target_id, size):
        node = get_by_id(target_id)
        old_next = node.next
        start_node, end_node = create_new_line(size)
        if old_next != None:
            end_node.next = old_next
            old_next.prev = end_node
        node.next = start_node
        start_node.prev = node

    def insert_prev(target_id, size):
        global root

        node = get_by_id(target_id)
        old_prev = node.prev
        start_node, end_node = create_new_line(size)
        if old_prev != None:
            start_node.prev = old_prev
            old_prev.next = start_node
        else:
            root = start_node
        node.prev = end_node
        end_node.next = node

    def process_query(q):
        if len(q) > 2:
            cmd, a, b = q
        else:
            cmd, a = q

        if cmd == 1:
            insert_next(a, b)
        elif cmd == 2:
            insert_prev(a, b)
        elif cmd == 3:
            node = get_by_id(a)
            if node.next == None or node.prev == None:
                print(-1)
                return
            print(node.prev.value, node.next.value)

    for q in queries:
        process_query(q)


q = int(sys.stdin.readline())
queries = [list(map(int, sys.stdin.readline().split())) for _ in range(q)]

sol(queries)

File: lim/연결리스트/Doubly_LinkedList_코드트리_유치원.java
package lim.연결리스트;

import java.util.*;
import java.io.*;

public class Doubly_LinkedList_코드트리_유치원 {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int q, nextNum = 1;
    static int[][] queries;
    static HashMap<Integer, Node> nodeByNum = new HashMap<>();

    static class Node {
        int num;
        Node prev, next;

        Node(int num) {
            this.num = num;
        }
    }

    static void insertPrev(Node target, Node newNode) {
        if (target.prev != null) {
            target.prev.next = newNode;
        }
        newNode.prev = target.prev;
        newNode.next = target;
        target.prev = newNode;
    }

    static void insertNext(Node target, Node newNode) {
        if (target.next != null) {
            target.next.prev = newNode;
        }
        newNode.next = target.next;
        newNode.prev = target;
        target.next = newNode;
    }

    static Node nextNode() {
        Node node = new Node(nextNum++);
        nodeByNum.put(node.num, node);
        return node;
    }

    static void sol() throws IOException {
        for (int[] query : queries) {
            int cmd = query[0];
            int val1 = query[1];
            int val2 = query[2];

            if (cmd == 1) {
                int count = val2;
                Node node = nodeByNum.get(val1);

                Node newNode = nextNode();
                insertNext(node, newNode);
                node = newNode;
                count--;

                while (count-- > 0) {
                    newNode = nextNode();
                    insertNext(node, newNode);
                    node = newNode;
                }
            } else if (cmd == 2) {
                int count = val2;
                Node node = nodeByNum.get(val1);

                Node newNode = nextNode();
                insertPrev(node, newNode);
                node = newNode;
                count--;

                while (count-- > 0) {
                    newNode = nextNode();
                    insertNext(node, newNode);
                    node = newNode;
                }
            } else if (cmd == 3) {
                Node node = nodeByNum.get(val1);
                if (node.prev == null || node.next == null) {
                    bw.write("-1\n");
                } else {
                    bw.write(node.prev.num + " ");
                    bw.write(node.next.num + "\n");
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        q = Integer.parseInt(st.nextToken());

        queries = new int[q][3];
        for (int i = 0; i < q; i++) {
            st = new StringTokenizer(br.readLine());
            queries[i][0] = Integer.parseInt(st.nextToken());
            queries[i][1] = Integer.parseInt(st.nextToken());
            if (queries[i][0] == 1 || queries[i][0] == 2) {
                queries[i][2] = Integer.parseInt(st.nextToken());
            }
        }

        nextNode();

        sol();
        br.close();
        bw.flush();
        bw.close();
    }
}